### PR TITLE
Pract.04: Add 10-12 | Miroslav Stoyanovski

### PR DESCRIPTION
**Pull request by mir4oni3**
Miroslav Stoyanovski
6 5MI0600195

### CODE REVIEW

File: SI/Sem.04/Pract.04/Task10.cpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#include <iostream>
+
+void toBinary(int n) {
+	int remainder = n % 2;
+	if (n < 1) {
+		return;
+	}
+	toBinary(n / 2);
```
Comment (GeorgiTerziev02) [inline SI/Sem.04/Pract.04/Task10.cpp:8]: Do it without recursion

File: SI/Sem.04/Pract.04/Task11.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+int min(int a, int b) {
+	return (a < b) ? a : b;
+}
+int min(int a, int b, int c, int d) {
+	return min(min(a, b), min(c, d));
+}
+int gcd(int a, int b) {
+	int highest = 1;
+	for (int i = 2; i <= min(a,b); i++) {
+		if (a % i == 0 && b % i == 0) {
```
Comment (GeorgiTerziev02) [inline SI/Sem.04/Pract.04/Task11.cpp:12]: You can optimize it with the Euclidean algorithm

File: SI/Sem.04/Pract.04/Task11.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+			highest = i;
+		}
+	}
+	return highest;
+}
+int gcd(int a, int b, int c, int d) {
+	int highest = 1;
+	for (int i = 2; i <= min(a, b, c, d); i++) {
+		if (a % i == 0 && b % i == 0 && c % i == 0 && d % i == 0) {
```
Comment (GeorgiTerziev02) [inline SI/Sem.04/Pract.04/Task11.cpp:21]: Find gcd of the first two, then of third and fourth, then gcd of the results as you did in min of four numbers

File: SI/Sem.04/Pract.04/Task12.cpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+		if (n % 2 == 1) {
+			return false;
+		}
+		n /= 2;
+	}
+	return false;
+}
+int nearestPowerOfTwo(int n) {
+	int left = n;
+	while (!isPowerOfTwo(left)) {
```
Comment (GeorgiTerziev02) [inline SI/Sem.04/Pract.04/Task12.cpp:17]: Can be optimized. This way you will check every number if it is power of two.
Example you enter 333. You will check all the numbers from 333 to 256 if they are power of two and then all the numbers from 333 to 512. Which is not optimal

File: SI/Sem.04/Pract.04/Task12.cpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	}
+	int right = n;
+	while (!isPowerOfTwo(right)) {
+		right++;
+	}
+
+	if (n - left <= right - n) {
+		return left;
+	}
+	return right;
```
Comment (GeorgiTerziev02) [inline SI/Sem.04/Pract.04/Task12.cpp:28]: ternary operator

**Review by GeorgiTerziev02 — CHANGES_REQUESTED** [review]

**Comment by Justsvetoslavov** [comment]
Closing due to merged solutions.

**Pull request closed without merging**